libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -80,29 +71,130 @@ bool KnowledgeBase::knownToBeZero(YulString _a)
 
 optional<u256> KnowledgeBase::valueIfKnownConstant(YulString _a)
 {
-	if (AssignedValue const* value = m_variableValues(_a))
-		if (Literal const* literal = get_if<Literal>(value->value))
-			return valueOfLiteral(*literal);
-	return {};
+	VariableOffset offset = explore(_a);
+	if (offset.reference == YulString{})
+		return offset.offset;
+	else
+		return nullopt;
+}
+
+optional<u256> KnowledgeBase::valueIfKnownConstant(Expression const& _expression)
+{
+	if (Identifier const* ident = get_if<Identifier>(&_expression))
+		return valueIfKnownConstant(ident->name);
+	else if (Literal const* lit = get_if<Literal>(&_expression))
+		return valueOfLiteral(*lit);
+	else
+		return {};
 }
 
-Expression KnowledgeBase::simplify(Expression _expression)
+KnowledgeBase::VariableOffset KnowledgeBase::explore(YulString _var)
 {
-	m_counter = 0;
-	return simplifyRecursively(std::move(_expression));
+	Expression const* value = nullptr;
+	if (m_valuesAreSSA)
+	{
+		// In SSA, a once determined offset is always valid, so we first see
+		// if we already computed it.
+		if (VariableOffset const* varOff = util::valueOrNullptr(m_offsets, _var))
+			return *varOff;
+		value = valueOf(_var);
+	}
+	else
+	{
+		// For non-SSA, we query the value first so that the variable is reset if it has changed
+		// since the last call.
+		value = valueOf(_var);
+		if (VariableOffset const* varOff = util::valueOrNullptr(m_offsets, _var))
+			return *varOff;
+	}
+
+	if (value)
```

Comment:
Can `valueOf` return a `nullptr`? If not, I don't think this check is necessary, as I don't see a way for us to reach this `if` without `value` being set.

Reply:
It returns `nullptr` for variables whose values are not movable, for example.